libkbfs/block_tree.go

```
@@ -291,159 +311,214 @@ func (bt *blockTree) getBlocksForOffsetRange(ctx context.Context,
-			if getDirect || childPtr.DirectType == IndirectBlock {
-				block, _, err := bt.getter(
-					groupCtx, bt.kmd, childPtr, bt.file, blockReadParallel)
-				if err != nil {
-					return err
-				}
 
-				// Recurse down to the level of the child.
-				pfr, blocks, nextBlockOffset, err = bt.getBlocksForOffsetRange(
-					groupCtx, childPtr, block, startOff, endOff, prefixOk,
-					getDirect)
-				if err != nil {
-					return err
-				}
-			} else {
-				// We don't care about direct blocks, so leave the
-				// `blocks` map `nil`.
-				pfr = [][]parentBlockAndChildIndex{nil}
-				nextBlockOffset = nil
+		childPath := make([]parentBlockAndChildIndex, len(job.pathPrefix)+1)
+		copy(childPath, job.pathPrefix)
+		childPath[len(childPath)-1] = parentBlockAndChildIndex{
+			pblock:     pblock,
+			childIndex: childIndex,
+		}
+
+		// We only need to fetch direct blocks if we've been asked
+		// to do so.  If the direct type of the pointer is
+		// unknown, we can assume all the children are direct
+		// blocks, since there weren't multiple levels of
+		// indirection before the introduction of the flag.
+		if job.getDirect || childPtr.DirectType == IndirectBlock {
+			// Create a task for the child, and either enqueue it or recurse to it.
+
+			subTask := task{
+				ptr:        childPtr,
+				pblock:     nil,
+				pathPrefix: childPath,
+				startOff:   job.startOff,
+				endOff:     job.endOff,
+				prefixOk:   job.prefixOk,
+				getDirect:  job.getDirect,
 			}
 
-			// Append self to the front of every path.
-			var r resp
-			for _, p := range pfr {
-				newPath := append([]parentBlockAndChildIndex{{
-					pblock:     pblock,
-					childIndex: childIndex,
-				}}, p...)
-				r.pathsFromRoot = append(r.pathsFromRoot, newPath)
+			// process the subtask by passing it to another goroutine; if they're all occupied, recurse instead.
+			select {
+			case jobs <- subTask:
+				wg.Add(1)
+			default:
+				bt.processGetBlocksTask(ctx, wg, subTask, jobs, results)
```

The error is dropped here, though if you implement my suggestion above and return it in `result` instead, you won't need to fix this.